### PR TITLE
Glowsticks, flares and lighting behaviours

### PR DESCRIPTION
**Pull request by nuke-makes-games**
This PR adds these things:

- **LightingBehaviourComponent** - It stores various lighting behaviour definitions and allows you to apply them to the point light component on the parent entity. 

- **ExpendableLightComponent** - Used on expendable things like glowsticks and flares to handle their various icon states, sound states, and lighting states.

Todo:

- [x] Make the expandable light component use a visualizer 
- [x] Refactor lighting behaviour to use a custom animator 
- [x] Attribute the sprites to me since i did em

Right now i have these functioning lighting behaviours:

- **Pulse** (the chosen property alternates between a min and max value using user's choice of nearest/linear/cubic).
- **Color cycle** (light cycles through a list of colors using nearest/linear/cubic).
- **Fade** (property interpolates from min to max value using nearest/linear/cubic).
- **Randomize** (property interpolates from current value to a random value chosen between min and max values using nearest/linear/cubic).

Things that would be nice to have:

- The ability to chain behaviours... So you can have a behaviour that segways into a different behaviour when it finishes. This is probably overkill though.

You have pretty granular control over how the light behaves with all of the various parameters. You can also have more than one of these behaviours defined in the prototype and you can toggle specific behaviours on and off using their given string ID (and even have several active at the same time). This means you have a pretty flexible system for giving a point light different states and behaviours that can be quickly managed via code and tweaked via yml for quick iterating.

Making a custom animator for this worked pretty well but the code could be tidier. There are some workarounds in place since having an animation that doesn't use keyframes isn't what the parent animators are made for.

Example of how to use it:

```
(a bunch of other components for whatever entity this is)
...
    - type: PointLight
      enabled: false
      color: red
      radius: 10.0
      energy: 10.0
    - type: LightBehaviour
      behaviours:
        - !type:RandomizeBehaviour # makes it bright and flickery like a campfire 
          id: flickery_mode
          interpolate: Nearest
          minDuration: 0.02
          maxDuration: 0.06
          startValue: 5.0
          endValue: 10.0
          property: Energy
          isLooped: true
        - !type:FadeBehaviour # shrinks the radius over the course of 10 seconds
          id: fade_out
          interpolate: Linear
          maxDuration: 10.0
          startValue: 10.0
          endValue: 2.0
          property: Radius
        - !type:ColorCycleBehaviour # alternating red and blue light, police car style
          id: get_down_on_the_fucking_ground
          interpolate: Nearest
          maxDuration: 0.5
          isLooped: true
          colors:
            - red
            - blue
```

And then to make use of these definitions from the YML you'd just do this on the clientside:

```
myLightBehaviourComp.StartLightBehaviour("get_down_on_the_fucking_ground"); // turn on the police lights

myLightBehaviourComp.StopLightBehaviour("get_down_on_the_fucking_ground"); // turn off the police lights

myLightBehaviourComp.StartLightBehaviour(); // turn on all 3 of the behaviours

myLightBehaviourComp.StopLightBehaviour(); // turn off all
```

Gifs made during the creation process (a bit old but you get the idea):

Toggle: https://cdn.discordapp.com/attachments/310555209753690112/747295854678769736/stre.gif
Random brightness: https://cdn.discordapp.com/attachments/310555209753690112/747295476906328124/rand.gif
Flicker: https://cdn.discordapp.com/attachments/310555209753690112/747294794983669770/flik.gif
Color sequence smoothed: https://cdn.discordapp.com/attachments/310555209753690112/747278967093985430/fdf.gif
Color sequence smoothed with rainbow colors: https://cdn.discordapp.com/attachments/310555209753690112/747280579824844800/ffff.gif
Using a random brightness behaviour on a flare: https://cdn.discordapp.com/attachments/310555209753690112/748025062640320622/test7.gif

### CODE REVIEW

**Comment by nuke-makes-games** [comment]
Alright this is good for review now. 🙏

File: Resources/Textures/Objects/Misc/flare.rsi/meta.json [hunk 1]
```diff
@@ -0,0 +1,126 @@
+{
+    "version": 1,
+    "size": {
+        "x": 32,
+        "y": 32
+    },
+    "license": "",
+    "copyright": "",
```
Comment (DrSmugleaf) [inline Resources/Textures/Objects/Misc/flare.rsi/meta.json:8]: These are missing a license and copyright. If you made them add CC-BY-SA-3.0 and your username.

File: Resources/Textures/Objects/Misc/glowstick.rsi/meta.json [hunk 2]
```diff
@@ -0,0 +1,121 @@
+{
+    "version": 1,
+    "size": {
+        "x": 32,
+        "y": 32
+    },
+    "license": "",
+    "copyright": "",
```
Comment (DrSmugleaf) [inline Resources/Textures/Objects/Misc/glowstick.rsi/meta.json:8]: Same here.

**Review by DrSmugleaf — CHANGES_REQUESTED** [review]

**Comment by nuke-makes-games** [comment]
Things i need to do:

- Make the expandable light component use a visualizer for the iconstates
- Refactor lighting behaviour to use a custom animator instead of this monstrosity it has become
- Attribute the sprites to me since i did em

**Comment by nuke-makes-games** [comment]
Submodule needs an update in order for the checks to pass. Otherwise i think i'm done making changes.

**Comment by SoulSloth** [comment]
Related to #1129

File: Resources/Textures/Objects/Misc/flare.rsi/meta.json [hunk 3]
```diff
@@ -0,0 +1,126 @@
+{
+    "version": 1,
+    "size": {
+        "x": 32,
+        "y": 32
+    },
+    "license": "CCBYNA3",
```
Comment (DrSmugleaf) [inline Resources/Textures/Objects/Misc/flare.rsi/meta.json:7]: I'm not sure this license exists, usually its either `CC-BY-SA-3.0`

File: Resources/Textures/Objects/Misc/glowstick.rsi/meta.json [hunk 4]
```diff
@@ -0,0 +1,121 @@
+{
+    "version": 1,
+    "size": {
+        "x": 32,
+        "y": 32
+    },
+    "license": "CCBYNA3",
```
Comment (DrSmugleaf) [inline Resources/Textures/Objects/Misc/glowstick.rsi/meta.json:7]: Same here.

**Review by DrSmugleaf — CHANGES_REQUESTED** [review]

File: Resources/Textures/Objects/Misc/flare.rsi/meta.json (hunk 3 above)
Comment (nuke-makes-games) [reply Resources/Textures/Objects/Misc/flare.rsi/meta.json:7]: hm, i copied that from another sprite so that's not good

**Review by DrSmugleaf — APPROVED** [review]

**Pull request merged**